lib/resty/jwt-validators.lua

```diff
@@ -0,0 +1,384 @@
+  ensure_is_type(check_val, "number", messages.wrong_type_validator, "number", "check_val")
+  return _M.opt_check(check_val, less_than_or_equal_function, "check_val", "number")
+end)
+
+
+--[[
+    A function to set the leeway (in seconds) used for is_not_before and is_not_expired.  The
+    default is to use 0 seconds
+]]--
+local system_leeway = 0
+function _M.set_system_leeway(leeway)
+  ensure_is_type(leeway, "number", "leeway must be a non-negative number")
+  ensure_is_non_negative(leeway, "leeway must be a non-negative number")
+  system_leeway = leeway
+end
+
+
+--[[
+    A function to set the system clock used for is_not_before and is_not_expired.  The
+    default is to use ngx.now
+]]--
+local system_clock = ngx.now
+function _M.set_system_clock(clock)
+  ensure_is_type(clock, "function", "clock must be a function")
+  
+  -- Check that clock returns the correct value
+  local t = clock()
+  ensure_is_type(t, "number", "clock function must return a non-negative number")
+  ensure_is_non_negative(t, "clock function must return a non-negative number")
+  system_clock = clock
+end
+
+-- Local helper function for date validation
+local function validate_is_date(val, claim, jwt_json)
+  ensure_is_non_negative(val, messages.wrong_type_claim, claim, "positive numeric value")
+  return true
+end
+
+-- Local helper for date formatting
+local function format_date_on_error(date_check_function, error_msg)
+  ensure_is_type(date_check_function, "function", messages.wrong_type_validator, "function", "date_check_function")
+  ensure_is_type(error_msg, "string", messages.wrong_type_validator, "string", error_msg)
+  return function(val, claim, jwt_json)
+    local ret = date_check_function(val, claim, jwt_json)
+    if ret == false then
+      error(string.format("'%s' claim %s %s", claim, error_msg, ngx.http_time(val)))
+    end
+    return true
+  end
+end
+
+--[[
+    Returns a validator that checks if the current time is not before the tested value
+    within the system's leeway.  This means that:
+      val <= (system_clock() + system_leeway).
+]]--
+define_validator("is_not_before", function()
+  return format_date_on_error(
+    _M.chain(validate_is_date, _M.opt_less_than_or_equal(system_clock() + system_leeway)),
+    "not valid until"
+  )
+end)
+
+
+--[[
+    Returns a validator that checks if the current time is not equal to or after the 
+    tested value within the system's leeway.  This means that:
+      val > (system_clock() - system_leeway).
+]]--
+define_validator("is_not_expired", function()
+  return format_date_on_error(
+    _M.chain(validate_is_date, _M.opt_greater_than(system_clock() - system_leeway)),
+    "expired at"
+  )
+end)
+
+
+--[[
+    Returns a validator that checks if the current time is the same as the tested value 
+    within the system's leeway.  This means that:
+      val >= (system_clock() - system_leeway) and val <= (system_clock() + system_leeway).
+]]--
+define_validator("is_at", function()
```

Comment:
> @fermaem 
> A jwt is supposed to last for some time, and be reused across many calls

This I believe is where the difference lies, I believe.  There is actually no requirement that jwt tokens are reused - and we use single-use tokens in my application.  We use `iat` (within the leeway) as a "quick" check for validity - and we also use `jti` to ensure that each token is only used once.

It would be the exact same thing as setting (on the client) `nbf=now()` and `exp=now()+1` and putting that in the token.  But it is smaller and more concise to use `iat`.

This check (server-side) *might* have been able to be replaced with _M.chain(_M.is_not_before(), _M.is_not_expired()) except for something is expired when it `equals` the date...meaning that the `chain` function will fail every time (because nothing is `>=now() and <now()`).

If it is not desired, I can remove this call...but it seems that it actually *does* fit a use case (mine :smile:).  Even if someone wanted to use it in validating a `my_custom_claim` claim, or something.  There is just a "hole" (`==`) in between `is_not_before` (`>=`) and `is_not_expired` (`<`) which I think could be filled.